feat(state): implement state management with useAppState hook and enhance documentation

odinr opened the pull request:
## 🎉 State Management for Fusion Framework

This PR introduces a complete state management solution for Fusion applications, making it easier to share persistent data between components while maintaining excellent developer experience.

### 🚀 Key Features

- **`useAppState` Hook**: React hook with `useState`-like API but with persistence and synchronization
- **Automatic Persistence**: State survives page refreshes using browser storage (PouchDB)
- **Real-time Synchronization**: Share state between components automatically
- **Type Safety**: Full TypeScript support with type inference
- **App-scoped Storage**: Automatic key prefixing to prevent collisions between apps
- **Optimistic Updates**: Lightning-fast UI with automatic error recovery

### 📦 Packages Changed

- **`@equinor/fusion-framework-app`** (minor): Added state module enabler and configuration
- **`@equinor/fusion-framework-react-app`** (major): Introduced `useAppState` hook and React integration
- **`@equinor/fusion-framework-app`** (patch): Improved AppConfigurator architecture and error handling

### 🛠️ Usage

**Enable state management:**
```typescript
import { enableState } from '@equinor/fusion-framework-app/enable-state';

export const configure = (configurator) => {
  enableState(configurator);
};


### Check off the following:
- [ ] Confirm that I checked changes to branch which I am merging into.
  - _I have validated included files_
  - _My code does not generate new linting warnings_
  - _My PR is not a duplicate, [check existing pr`s](https://github.com/equinor/fusion-framework/pulls)_
- [ ] Confirm that the I have completed the [self-review checklist](https://github.com/equinor/fusion-framework/blob/main/contributing/self-review.md).

- [ ] Confirm that my changes meet our [code of conduct](https://github.com/equinor/fusion-framework/blob/main/CODE_OF_CONDUCT.md).

odinr committed:
feat(module-state): add robust storage interface, PouchDB adapter, and comprehensive tests

- Implement generic IStorage interface and types for state storage
- Add PouchDbStorage adapter with live change tracking and bulk operations
- Provide update-storage-item(s) RxJS operators for reactive state updates
- Add full test suite for StateProvider and storage operators
- Remove legacy storage adapters and refactor exports for clarity
- Update dependencies and TypeScript config for new module structure

odinr committed:
feat(app): add state management module with PouchDB support and enhance AppConfigurator architecture

odinr committed:
feat(app_state): implement state management with useAppState hook and enhance documentation

odinr committed:
feat(cookbook-react-state): add comprehensive state management cookbook with useAppState examples and documentation

Noggling commented on packages/app/src/enable-state.ts:
```diff
@@ -0,0 +1,56 @@
+  enableStateModule(configurator, (config) => {
+    // Use PouchDB with app-specific key prefix to ensure state isolation
+    // The key_prefix prevents state collision between different apps using the same database
+    const storage = new PouchDbStorage(APP_DB_NAME, { key_prefix: configurator.manifest.appKey });
```
should we add:

            /**
             * How many old revisions we keep track (not a copy) of.
             */
            revs_limit?: number | undefined;
            /**
             * Size of the database (Most significant for Safari)
             * option to set the max size in MB that Safari will grant to the local database. Valid options are: 10, 50, 100, 500 and 1000
             * ex_ new PouchDB("dbName", {size:100});
             */
            size?: number | undefined;

odinr replied:
the object now takes the db option object, so all of em can be configured

alftore replied:
Done in e6de8665. Added `revs_limit` and `size` explicitly to `PouchDbStorageOptions` with JSDoc. Both exist on `PouchDB.Configuration.LocalDatabaseConfiguration` but the union type makes them non-obvious without narrowing — the explicit declaration surfaces them for discoverability and documents the valid `size` values (10/50/100/500/1000 MB) including the Safari note.

odinr committed:
feat(cookbooks-app-react-state-replication): Create a cookbook for state replication

- Added TodoListItem component for individual todo items.
- Created TodoListManager to manage the todo list state and handle CRUD operations.
- Introduced TodoStats component to display statistics of the todo items.
- Developed types for TodoItem and TodoList for better type safety.
- Configured PouchDB for local and remote CouchDB replication.
- Implemented event handling for CouchDB sync events.
- Created Home, Profile, and Todo pages to structure the application.
- Added routing and navigation using react-router.
- Included SyncStatusMonitor to display sync status in the UI.
- Updated package dependencies and added TypeScript configurations.

odinr committed:
docs(state): add cautionary notes and best practices for state management

odinr committed:
feat: add state management module with sync capabilities and React hooks

- Add @equinor/fusion-framework-module-state with PouchDB sync storage
- Add state management support to @equinor/fusion-framework-app
- Add useAppState hook to @equinor/fusion-framework-react-app
- Add state management cookbooks for examples and replication demos
- Reorganize event system with comprehensive state event types
- Update dependencies and fix formatting issues

alftore committed:
fix(state): address code quality review comments on PR #3286

- remove unused ProfileManager import in app-react-state-replication/App.tsx
- remove unused AppModulesInitializedEvent import in AppConfigurator.ts
- remove unused 'type' from event destructuring in StateProvider.ts
- export PouchDbSyncStorage class
- add revs_limit and size options to PouchDbStorageOptions

alftore commented:
Added `revs_limit` and `size` to `PouchDbStorageOptions` with JSDoc in e6de8665. Both are already present on `PouchDB.Configuration.LocalDatabaseConfiguration` (which `DatabaseConfiguration` unions over), but the explicit declaration makes them discoverable without needing to narrow the union type, and the valid `size` values (10/50/100/500/1000 MB) are now documented inline including the Safari note.

alftore commented on packages/modules/state/src/storage/PouchDbSyncStorage.ts:
```diff
@@ -0,0 +1,83 @@
+  syncOptions: PouchDB.Replication.SyncOptions;
+};
+
+class PouchDbSyncStorage extends PouchDbStorage {
```
Addressed in e6de8665 — class kept private (intentionally internal, no tests and no current consumers). Added a `biome-ignore lint/correctness/noUnusedVariables` suppression with explicit rationale rather than exporting an untested API.

odinr committed:
fix: sync pnpm-lock.yaml with workspace:^ specifiers

odinr committed:
fix: repair rebase-induced regressions and TS7 build breakage

- restore missing return statement in configureModules (silently dropped by git's 3-way merge during rebase)
- fix broken absolute-style import in state-sync-observer.ts
- drop unsupported baseUrl from state module tsconfig (removed in TypeScript 7)
- pin explicit generic types on createModuleConfigs' reduce operator to fix unknown-type inference under TS7 when built via composite project references

odinr committed:
chore: fix fusion-lint diagnostics across state module, app, react/app, and cookbook

- add missing TSDoc and intent comments
- split multi-export files into single-export files with barrel exports
- rename files to match their default export per fusion-lint filename-convention
- remove TODO comments lacking a tracking issue reference

odinr committed:
fix: replace react-router-dom with @equinor/fusion-framework-react-router in state-replication cookbook

The cookbook mixed react-router-dom's RouterProvider with the router
instance returned by useRouter (@equinor/fusion-framework-react-app),
which internally uses @remix-run/router via the navigation module.
The mismatched react-router runtimes caused a crash in useRoutesImpl
at render time. Also clean up remaining biome lint issues (unused
imports/vars, array index keys, banned {} type) left over from the
earlier fusion-lint cleanup pass.

odinr committed:
docs: sync CODEMAP.md with module-state package and new state cookbooks

odinr commented on packages/modules/state/src/events/index.ts:
```diff
@@ -0,0 +1,95 @@
+  | StateSyncCompleteEvent<T>
+  | StateSyncErrorEvent
+  | StateSyncStatusEvent
+  | StateErrorEvent;
```
Agreed — confirmed `observePouchDbSync` never emits `StateErrorEvent` (only `StateSyncErrorEvent` via `StateSyncEvent.Error`), so it doesn't belong in the sync union. Removed in e8252157a.

odinr commented on packages/react/app/src/state/useStateSyncEvents.ts:
```diff
@@ -0,0 +1,59 @@
+      .subscribe((event) => {
+        const next = [...event$.getValue(), event];
+        event$.next(next.length > limitRef.current ? next.slice(-limitRef.current) : next);
+      });
```
Confirmed — `slice(-0)` is a no-op, unlike `slice(-1)`, so `limit=0` retained everything. Clamped to 0 and special-cased it to return `[]`. Fixed in 6da04e606.

odinr commented on packages/app/README.md:
````diff
@@ -16,6 +27,32 @@ pnpm add @equinor/fusion-framework-app
 ## Quick Start
+```typescript
+import type { AppModuleInitiator } from '@equinor/fusion-framework-app';
+import { enableState } from '@equinor/fusion-framework-app/state';
````
Confirmed the package only exports the enabler under `./enable-state` — fixed the example import in ca5f5bea4.

odinr commented on packages/modules/state/package.json:
```diff
@@ -0,0 +1,75 @@
+  "types": "./dist/types/index.d.ts",
+  "exports": {
+    ".": {
+      "import": "./dist/esm/index.js",
+      "types": "./dist/types/index.d.ts"
+    },
+    "./events": {
+      "import": "./dist/esm/events/index.js",
+      "types": "./dist/types/storeventsage/index.d.ts"
+    },
+    "./storage": {
+      "import": "./dist/esm/storage/index.js",
+      "types": "./dist/types/storage/index.d.ts"
+    },
+    "./package.json": "./package.json",
+    "./README.md": "./README.md"
+  },
+  "typesVersions": {
+    "*": {
+      ".": [
+        "dist/types/index.d.ts"
+      ],
+      "storage": [
+        "dist/types/storage/index.d.ts"
+      ]
+    }
+  },
```
Fixed the typo'd path and also added the missing `events` `typesVersions` entry (same class of bug) in 978c0e721.

odinr commented on packages/react/app/src/state/useAppState.ts:
```diff
@@ -0,0 +1,298 @@
+          });
+      }
+    },
+    [stateProvider, value$],
```
Added `getValue` to the `useCallback` dependency list. Fixed in 849b9a61e.

odinr commented on packages/modules/state/src/storage/PouchDbSyncStorage.ts:
```diff
@@ -0,0 +1,122 @@
+      retry, // Automatically retry on failures
+      heartbeat, // Send heartbeat to detect connection issues
+      timeout, // Connection timeout
+    });
```
Good catch. Now spreading caller `options` first so only `live`/`retry`/`heartbeat`/`timeout` are defaulted and every other `SyncOptions` field passes through unmodified. Fixed in dc98e88ab.

odinr commented on packages/app/src/AppConfigurator.ts:
```diff
@@ -98,9 +114,9 @@ export interface IAppConfigurator<
  * helper methods for HTTP client and service-discovery setup.
  *
- * `AppConfigurator` is created internally by {@link configureModules}. It registers
+ * `AppConfigurator` is created internally by {\@link configureModules}. It registers
```
Fixed the escaped `@` so the tag renders as a real `{@link}`. Fixed in 1cd6ca1ae.

odinr commented on packages/modules/module/src/lib/configurator/phases/create-module-configs.ts:
```diff
@@ -95,7 +95,7 @@ export async function createModuleConfigs<TRef>(
       }
     }),
-    reduce(
+    reduce<Record<string, any>, Record<string, any>>(
```
Switched to `unknown` — same inference fix without introducing `any`. Fixed in 300acc677.

odinr commented on packages/react/app/src/state/useAppState.ts:
```diff
@@ -0,0 +1,298 @@
+    (action: SetStateAction<T | undefined>) => {
+      // Handle both direct values and updater functions (like React's useState)
+      const value = typeof action === 'function' ? action(value$.value) : action;
+
```
Good catch — `setValue`'s updater now receives `getValue(value$.value)` (same defaulting the snapshot uses) instead of the raw, possibly-`undefined` value. Fixed in 849b9a61e.

odinr commented on packages/modules/state/src/types.ts:
```diff
@@ -0,0 +1,21 @@
+export type AllowedValue =
+  | string
+  | Array<unknown>
+  | Record<string | symbol | number, unknown>
+  | Set<unknown>
+  | number
+  | boolean
+  | undefined
+  | null;
```
Agreed — dropped `Set` and symbol/number record keys from `AllowedValue` so it matches what the PouchDB-backed storage can actually round-trip; updated the README's supported-types list to match. Fixed in e79bc2d41.

odinr commented on packages/react/app/src/state/useAppState.ts:
```diff
@@ -0,0 +1,298 @@
+    return () => {
+      subscription.unsubscribe();
+    };
+  }, [stateProvider.observeItem, value$]);
```
Agreed, `observeItem` is a stable prototype method so the effect wouldn't re-run on a provider swap. Changed the dependency to `stateProvider`. Fixed in 849b9a61e.

odinr commented on packages/modules/state/src/storage/Storage.interface.ts:
```diff
@@ -0,0 +1,107 @@
+import type { Observable } from 'rxjs';
+
+import type { AllowedValue, StateEventType } from '@equinor/fusion-framework-module-state';
+
+import type { StorageItem, StorageResult } from './types.js';
+
```
Agreed — switched `AllowedValue` to `../types.js` and `StateEventType` to `../events/index.js`. Fixed in 75d2b28b1 (same commit fixes the other instances of this pattern in the package).

odinr commented on packages/modules/state/src/events/StateEntryCreatedEvent.ts:
```diff
@@ -0,0 +1,74 @@
+
+import { FrameworkEvent, type FrameworkEventInit } from '@equinor/fusion-framework-module-event';
+import type { AllowedValue } from '@equinor/fusion-framework-module-state';
+import type { StorageItem } from '../storage/types.js';
```
Fixed, along with the same self-import pattern elsewhere in the package. Fixed in 75d2b28b1.

odinr commented on packages/modules/state/src/StateModuleConfig.ts:
```diff
@@ -0,0 +1,22 @@
+ * Zod schema for the IStorage interface.
+ *
+ * @see {@link IStorage} - `@equinor/fusion-framework-module-state/storage`
+ */
```
Confirmed — that trailing doc block described a validator that was never implemented, plus a stray `@see` pointing at a nonexistent Zod schema. Removed both in ffb1752d1.

odinr commented on packages/modules/state/src/storage/types.ts:
```diff
@@ -0,0 +1,80 @@
+import type { AllowedValue } from '@equinor/fusion-framework-module-state';
```
Good catch. Switched to a relative import (`../types.js`), and fixed the same self-import bug in 6 other files in this package that had it (`Storage.interface.ts`, `PouchDbStorage.ts`, the three `StateEntry*Event.ts` files, and `storage.mock.ts`). Fixed in 75d2b28b1.
